### PR TITLE
Add support for PancakeSwap V3

### PR DESCRIPTION
**Pull request by gnarlycow**

### CODE REVIEW

File: contracts/PCSV3PositionUtils.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import {ERC20Callee} from "./libraries/ERC20Caller.sol";
+import {PoolUtils} from "./PoolUtils.sol";
+
+struct PCSV3Slot0 {
+    uint160 sqrtPriceX96;
+    int24 tick;
+    uint16 observationIndex;
+    uint16 observationCardinality;
+    uint16 observationCardinalityNext;
+    uint32 feeProtocol;
```
Comment (shuhuiluo) [inline contracts/PCSV3PositionUtils.sol]: Can change the original `Slot0. feeProtocol` to `uint32` and reduce a lot of duplicates.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: contracts/EphemeralPoolSlots.sol [hunk 2]
```diff
@@ -72,7 +73,89 @@ contract EphemeralPoolSlots is PoolUtils {
...
+                    uint32 feeProtocol,
+                    bool unlocked
+                ) = IPancakeV3Pool(V3PoolCallee.unwrap(pool)).slot0();
+                // PancakeV3Pool's slot0() fields actually span two storage slots (slot 0 and 1) as a result of their changing `feeProtocol` from uint8 to uint32.
+                // The first 5 fields are packed into slot 0, and the last 2 fields (`feeProtocol` and `unlocked`) are packed into slot 1.
+                // See https://evm.storage/eth/19541394/0x6ca298d2983ab03aa1da7679389d955a4efee15c/slot0#map for a visual representation.
+                uint256 slot0;
+                uint256 slot1;
+                assembly {
+                    slot0 := or(shl(216, observationCardinalityNext), slot0)
```
Comment (shuhuiluo) [inline contracts/EphemeralPoolSlots.sol]:
```suggestion
                    slot0 := shl(216, observationCardinalityNext)
```

File: test/foundry/PositionLens.t.sol [hunk 3]
```diff
@@ -84,6 +88,25 @@ contract PositionLensTest is BaseTest {
...
+            int24 tick,
+            uint16 observationIndex,
+            uint16 observationCardinality,
+            uint16 observationCardinalityNext,
+            uint32 feeProtocol,
+            bool unlocked
+        )
+    {
+        return IUniswapV3Pool(pool).slot0();
+    }
```
Comment (shuhuiluo) [inline test/foundry/PositionLens.t.sol]:
```suggestion
```

File: test/foundry/PositionLens.t.sol [hunk 4]
```diff
@@ -96,35 +119,45 @@ contract PositionLensTest is BaseTest {
...
-            (uint160 sqrtPriceX96, int24 tick, , , , , ) = pool.slot0();
+            (
+                uint160 sqrtPriceX96,
+                int24 tick,
+                uint16 observationIndex,
+                uint16 observationCardinality,
+                uint16 observationCardinalityNext,
+                uint32 feeProtocol,
+                bool unlocked
+            ) = getPoolSlot0(pool);
```
Comment (shuhuiluo) [inline test/foundry/PositionLens.t.sol]:
```suggestion
            ) = IPancakeV3Pool(pool).slot0();
```

File: test/foundry/PositionLens.t.sol [hunk 5]
```diff
@@ -151,3 +184,39 @@ contract PositionLensTest is BaseTest {
...
+            int24 tick,
+            uint16 observationIndex,
+            uint16 observationCardinality,
+            uint16 observationCardinalityNext,
+            uint32 feeProtocol,
+            bool unlocked
+        )
+    {
+        return IPancakeV3Pool(pool).slot0();
+    }
```
Comment (shuhuiluo) [inline test/foundry/PositionLens.t.sol]:
```suggestion
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: package.json [hunk 6]
```diff
@@ -47,18 +47,18 @@
     "typechain": "hardhat typechain"
   },
   "dependencies": {
+    "@aperture_finance/uniswap-v3-automation-sdk": "3.0.0-alpha",
```
Comment (shuhuiluo) [inline package.json]: Circular dependency.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

**Review by gaohan137 — APPROVED** [review]

**Pull request merged**